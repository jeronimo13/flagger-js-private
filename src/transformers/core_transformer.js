import {logger} from '../logger'
import {isValidFlagConfig} from '../validators/core_validator'
import Population, {
  RULE_TYPE_STRING,
  RULE_TYPE_INT,
  RULE_TYPE_FLOAT,
  RULE_TYPE_BOOLEAN,
  RULE_TYPE_DATE,
  RULE_TYPE_DATETIME,
  RULE_OPERATOR_IS,
  RULE_OPERATOR_IS_NOT,
  RULE_OPERATOR_IN,
  RULE_OPERATOR_NOT_IN,
  RULE_OPERATOR_LT,
  RULE_OPERATOR_LTE,
  RULE_OPERATOR_GT,
  RULE_OPERATOR_GTE,
  RULE_OPERATOR_FROM,
  RULE_OPERATOR_UNTIL,
  RULE_OPERATOR_AFTER,
  RULE_OPERATOR_BEFORE
} from '../population'

const RULE_TYPE_TO_ALLOWED_OPERATORS = {
  [RULE_TYPE_STRING]: new Set([
    RULE_OPERATOR_IS,
    RULE_OPERATOR_IS_NOT,
    RULE_OPERATOR_IN,
    RULE_OPERATOR_NOT_IN
  ]),
  [RULE_TYPE_INT]: new Set([
    RULE_OPERATOR_IS,
    RULE_OPERATOR_IS_NOT,
    RULE_OPERATOR_IN,
    RULE_OPERATOR_NOT_IN,
    RULE_OPERATOR_LT,
    RULE_OPERATOR_LTE,
    RULE_OPERATOR_GT,
    RULE_OPERATOR_GTE
  ]),
  [RULE_TYPE_FLOAT]: new Set([
    RULE_OPERATOR_IS,
    RULE_OPERATOR_IS_NOT,
    RULE_OPERATOR_IN,
    RULE_OPERATOR_NOT_IN,
    RULE_OPERATOR_LT,
    RULE_OPERATOR_LTE,
    RULE_OPERATOR_GT,
    RULE_OPERATOR_GTE
  ]),
  [RULE_TYPE_BOOLEAN]: new Set([RULE_OPERATOR_IS, RULE_OPERATOR_IS_NOT]),
  [RULE_TYPE_DATE]: new Set([
    RULE_OPERATOR_IS,
    RULE_OPERATOR_IS_NOT,
    RULE_OPERATOR_IN,
    RULE_OPERATOR_NOT_IN,
    RULE_OPERATOR_FROM,
    RULE_OPERATOR_UNTIL,
    RULE_OPERATOR_AFTER,
    RULE_OPERATOR_BEFORE
  ]),
  [RULE_TYPE_DATETIME]: new Set([
    RULE_OPERATOR_IS,
    RULE_OPERATOR_IS_NOT,
    RULE_OPERATOR_IN,
    RULE_OPERATOR_NOT_IN,
    RULE_OPERATOR_FROM,
    RULE_OPERATOR_UNTIL,
    RULE_OPERATOR_AFTER,
    RULE_OPERATOR_BEFORE
  ])
}

export const transformFlagConfig = flagConfig => {
  if (!isValidFlagConfig(flagConfig)) {
    return null
  }

  const flags = []
  const keys = Object.keys(flagConfig)
  for (let i = 0; i < keys.length; i++) {
    const key = keys[i]
    const config = flagConfig[key]
    const active = config.active !== undefined ? config.active : true
    const flagInfo = {
      flagType: 'basic',
      hashKey: key,
      isPaused: !active,
      isWebAccessible: __BROWSER__, // eslint-disable-line no-undef
      codename: key,
      flagStatus: 'operational'
    }
    const whitelist = config.whitelist || []
    const blacklist = config.blacklist || []
    const blacklistSet = new Set(blacklist)
    const filteredWhitelist = whitelist.filter(i => !blacklistSet.has(i))

    flagInfo.treatments = [
      {
        treatmentId: 'off-treatment',
        codename: 'off',
        isControl: false,
        isOffTreatment: true
      },
      {
        treatmentId: 'on-treatment',
        codename: 'on',
        isControl: false,
        isOffTreatment: false
      }
    ]

    flagInfo.overrides = []

    flagInfo.overrides = flagInfo.overrides.concat(
      filteredWhitelist.map(i => ({
        treatmentId: 'on-treatment',
        entityType: 'User',
        entityId: i.toString()
      }))
    )

    flagInfo.overrides = flagInfo.overrides.concat(
      blacklist.map(i => ({
        treatmentId: 'off-treatment',
        entityType: 'User',
        entityId: i.toString()
      }))
    )

    flagInfo.splits = [{treatmentId: 'on-treatment', percentage: 1}]

    if (config.population) {
      const rules = config.population || []
      const ruleInfos = []

      for (let j = 0; j < rules.length; j++) {
        const r = rules[j]
        let type
        if (Array.isArray(r.value)) {
          const types = new Set(
            r.value.map(v => Population.categorizeValueType(v))
          )

          if (types.size != 1) {
            logger(
              "Population's filter criteria each should have a singular value type. In other words, do not mix strings with numbers in the same array, for example."
            )
            return null
          }

          type = Array.from(types)[0]
        } else {
          type = Population.categorizeValueType(r.value)
        }

        const rInfo = {
          attributeName: r.attribute,
          attributeType: type,
          operator: r.operator,
          value: !Array.isArray(r.value) ? r.value : null,
          valueList: !Array.isArray(r.value) ? null : r.value
        }

        if (!RULE_TYPE_TO_ALLOWED_OPERATORS[type].has(r.operator)) {
          logger(
            `Population's filter operator \`${
              r.operator
            }\` is not allowed for filter type \`${type}\``
          )
          return null
        }

        if (rInfo.valueList) {
          if (
            rInfo.operator !== RULE_OPERATOR_IN &&
            rInfo.operator !== RULE_OPERATOR_NOT_IN
          ) {
            logger(
              "Population's filter operator must be `in` or `not_in` if the value is an array."
            )
            return null
          }
        } else {
          if (
            rInfo.operator === RULE_OPERATOR_IN &&
            rInfo.operator === RULE_OPERATOR_NOT_IN
          ) {
            logger(
              "Population's filter operator must not be `in` or `not_in` if the value is a number, boolean or string."
            )
            return null
          }
        }

        ruleInfos.push(rInfo)
      }

      flagInfo.populations = [
        {
          hashKey: 'population-1',
          entityType: 'User',
          percentage: config.sample || 0.0,
          rules: ruleInfos,
          universes: []
        }
      ]
    } else {
      flagInfo.populations = [
        {
          hashKey: 'population-1',
          entityType: 'User',
          percentage: config.sample || 0.0,
          rules: [],
          universes: []
        }
      ]
    }

    flags.push(flagInfo)
  }

  return {
    flags: flags,
    env: {
      hashKey: 'env-1',
      envKey: null
    }
  }
}
