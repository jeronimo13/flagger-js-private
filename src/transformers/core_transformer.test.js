import {transformFlagConfig} from './core_transformer'

test('can transform valid flagConfig because {} is valid', () => {
  expect(transformFlagConfig({})).toMatchSnapshot()
})

test('can transform valid flagConfig because given flagConfig is valid', () => {
  const flagConfig = {
    'super-rare-pokemon': {
      population: [
        {attribute: 'country', operator: 'is', value: 'Canada'},
        {attribute: 'age', operator: 'lt', value: 15}
      ],
      active: true
    },
    trix: {
      population: [
        {attribute: 'birthDate', operator: 'after', value: '2000-09-01'}
      ],
      sample: 1.0
    },
    'new-dashboard': {
      sample: 0.15
    },
    'modern-warfare-xxi': {
      whitelist: [54, 64, 23, 65, 59],
      blacklist: [43, 34]
    },
    'more-ads': {
      population: [
        {
          attribute: 'country',
          operator: 'in',
          value: ['USA', 'CANADA', 'DUBAI']
        },
        {attribute: 'age', operator: 'lt', value: 40}
      ],
      sample: 0.3,
      whitelist: [45, 34, 84],
      blacklist: [43, 35, 65],
      active: false
    }
  }
  expect(transformFlagConfig(flagConfig)).toMatchSnapshot()
})

test('cannot transform flagConfig because value is not compatible with operator', () => {
  const flagConfig = {
    'super-rare-pokemon': {
      population: [{attribute: 'country', operator: 'is', value: [1, 'Canada']}]
    }
  }
  expect(transformFlagConfig(flagConfig)).toEqual(null)
})

test('cannot transform flagConfig because of mixed value types', () => {
  const flagConfig = {
    'super-rare-pokemon': {
      population: [{attribute: 'country', operator: 'in', value: [1, 'Canada']}]
    }
  }
  expect(transformFlagConfig(flagConfig)).toEqual(null)
})

test('cannot transform flagConfig because of empty array', () => {
  const flagConfig = {
    'super-rare-pokemon': {
      population: [{attribute: 'country', operator: 'in', value: []}]
    }
  }
  expect(transformFlagConfig(flagConfig)).toEqual(null)
})

test('cannot transform flagConfig because of incompatible operator and type', () => {
  const flagConfig = {
    'robinhood-free': {
      population: [
        {attribute: 'dateSignedUp', operator: 'lte', value: '2018-01-01'}
      ]
    }
  }
  expect(transformFlagConfig(flagConfig)).toEqual(null)
})
