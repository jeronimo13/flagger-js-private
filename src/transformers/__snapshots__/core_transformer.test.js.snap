// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can transform valid flagConfig because {} is valid 1`] = `
Object {
  "env": Object {
    "envKey": null,
    "hashKey": "env-1",
  },
  "flags": Array [],
}
`;

exports[`can transform valid flagConfig because given flagConfig is valid 1`] = `
Object {
  "env": Object {
    "envKey": null,
    "hashKey": "env-1",
  },
  "flags": Array [
    Object {
      "codename": "super-rare-pokemon",
      "flagStatus": "operational",
      "flagType": "basic",
      "hashKey": "super-rare-pokemon",
      "isPaused": false,
      "isWebAccessible": true,
      "overrides": Array [],
      "populations": Array [
        Object {
          "entityType": "User",
          "hashKey": "population-1",
          "percentage": 0,
          "rules": Array [
            Object {
              "attributeName": "country",
              "attributeType": "string",
              "operator": "is",
              "value": "Canada",
              "valueList": null,
            },
            Object {
              "attributeName": "age",
              "attributeType": "int",
              "operator": "lt",
              "value": 15,
              "valueList": null,
            },
          ],
          "universes": Array [],
        },
      ],
      "splits": Array [
        Object {
          "percentage": 1,
          "treatmentId": "on-treatment",
        },
      ],
      "treatments": Array [
        Object {
          "codename": "off",
          "isControl": false,
          "isOffTreatment": true,
          "treatmentId": "off-treatment",
        },
        Object {
          "codename": "on",
          "isControl": false,
          "isOffTreatment": false,
          "treatmentId": "on-treatment",
        },
      ],
    },
    Object {
      "codename": "trix",
      "flagStatus": "operational",
      "flagType": "basic",
      "hashKey": "trix",
      "isPaused": false,
      "isWebAccessible": true,
      "overrides": Array [],
      "populations": Array [
        Object {
          "entityType": "User",
          "hashKey": "population-1",
          "percentage": 1,
          "rules": Array [
            Object {
              "attributeName": "birthDate",
              "attributeType": "date",
              "operator": "after",
              "value": "2000-09-01",
              "valueList": null,
            },
          ],
          "universes": Array [],
        },
      ],
      "splits": Array [
        Object {
          "percentage": 1,
          "treatmentId": "on-treatment",
        },
      ],
      "treatments": Array [
        Object {
          "codename": "off",
          "isControl": false,
          "isOffTreatment": true,
          "treatmentId": "off-treatment",
        },
        Object {
          "codename": "on",
          "isControl": false,
          "isOffTreatment": false,
          "treatmentId": "on-treatment",
        },
      ],
    },
    Object {
      "codename": "new-dashboard",
      "flagStatus": "operational",
      "flagType": "basic",
      "hashKey": "new-dashboard",
      "isPaused": false,
      "isWebAccessible": true,
      "overrides": Array [],
      "populations": Array [
        Object {
          "entityType": "User",
          "hashKey": "population-1",
          "percentage": 0.15,
          "rules": Array [],
          "universes": Array [],
        },
      ],
      "splits": Array [
        Object {
          "percentage": 1,
          "treatmentId": "on-treatment",
        },
      ],
      "treatments": Array [
        Object {
          "codename": "off",
          "isControl": false,
          "isOffTreatment": true,
          "treatmentId": "off-treatment",
        },
        Object {
          "codename": "on",
          "isControl": false,
          "isOffTreatment": false,
          "treatmentId": "on-treatment",
        },
      ],
    },
    Object {
      "codename": "modern-warfare-xxi",
      "flagStatus": "operational",
      "flagType": "basic",
      "hashKey": "modern-warfare-xxi",
      "isPaused": false,
      "isWebAccessible": true,
      "overrides": Array [
        Object {
          "entityId": "54",
          "entityType": "User",
          "treatmentId": "on-treatment",
        },
        Object {
          "entityId": "64",
          "entityType": "User",
          "treatmentId": "on-treatment",
        },
        Object {
          "entityId": "23",
          "entityType": "User",
          "treatmentId": "on-treatment",
        },
        Object {
          "entityId": "65",
          "entityType": "User",
          "treatmentId": "on-treatment",
        },
        Object {
          "entityId": "59",
          "entityType": "User",
          "treatmentId": "on-treatment",
        },
        Object {
          "entityId": "43",
          "entityType": "User",
          "treatmentId": "off-treatment",
        },
        Object {
          "entityId": "34",
          "entityType": "User",
          "treatmentId": "off-treatment",
        },
      ],
      "populations": Array [
        Object {
          "entityType": "User",
          "hashKey": "population-1",
          "percentage": 0,
          "rules": Array [],
          "universes": Array [],
        },
      ],
      "splits": Array [
        Object {
          "percentage": 1,
          "treatmentId": "on-treatment",
        },
      ],
      "treatments": Array [
        Object {
          "codename": "off",
          "isControl": false,
          "isOffTreatment": true,
          "treatmentId": "off-treatment",
        },
        Object {
          "codename": "on",
          "isControl": false,
          "isOffTreatment": false,
          "treatmentId": "on-treatment",
        },
      ],
    },
    Object {
      "codename": "more-ads",
      "flagStatus": "operational",
      "flagType": "basic",
      "hashKey": "more-ads",
      "isPaused": true,
      "isWebAccessible": true,
      "overrides": Array [
        Object {
          "entityId": "45",
          "entityType": "User",
          "treatmentId": "on-treatment",
        },
        Object {
          "entityId": "34",
          "entityType": "User",
          "treatmentId": "on-treatment",
        },
        Object {
          "entityId": "84",
          "entityType": "User",
          "treatmentId": "on-treatment",
        },
        Object {
          "entityId": "43",
          "entityType": "User",
          "treatmentId": "off-treatment",
        },
        Object {
          "entityId": "35",
          "entityType": "User",
          "treatmentId": "off-treatment",
        },
        Object {
          "entityId": "65",
          "entityType": "User",
          "treatmentId": "off-treatment",
        },
      ],
      "populations": Array [
        Object {
          "entityType": "User",
          "hashKey": "population-1",
          "percentage": 0.3,
          "rules": Array [
            Object {
              "attributeName": "country",
              "attributeType": "string",
              "operator": "in",
              "value": null,
              "valueList": Array [
                "USA",
                "CANADA",
                "DUBAI",
              ],
            },
            Object {
              "attributeName": "age",
              "attributeType": "int",
              "operator": "lt",
              "value": 40,
              "valueList": null,
            },
          ],
          "universes": Array [],
        },
      ],
      "splits": Array [
        Object {
          "percentage": 1,
          "treatmentId": "on-treatment",
        },
      ],
      "treatments": Array [
        Object {
          "codename": "off",
          "isControl": false,
          "isOffTreatment": true,
          "treatmentId": "off-treatment",
        },
        Object {
          "codename": "on",
          "isControl": false,
          "isOffTreatment": false,
          "treatmentId": "on-treatment",
        },
      ],
    },
  ],
}
`;
